transport_plugins/bled112/iotile_transport_bled112/bled112.py

```diff
@@ -637,22 +663,57 @@ def _parse_v2_advertisement(self, rssi, sender, data):
         #   bit 5: Encryption key is user key
         #   bit 6: broadcast data is time synchronized to avoid leaking
         #   information about when it changes
+        is_pending_data = bool(flags & (1 << 0))
+        is_low_voltage = bool(flags & (1 << 1))
+        is_user_connected = bool(flags & (1 << 2))
+        is_encrypted = bool(flags & (1 << 3))
+        is_device_key = bool(flags & (1 << 4))
+        is_user_key = bool(flags & (1 << 5))
 
         self._device_scan_counts.setdefault(device_id, {'v1': 0, 'v2': 0})['v2'] += 1
 
         info = {'connection_string': sender,
                 'uuid': device_id,
-                'pending_data': bool(flags & (1 << 0)),
-                'low_voltage': bool(flags & (1 << 1)),
-                'user_connected': bool(flags & (1 << 2)),
+                'pending_data': is_pending_data,
+                'low_voltage': is_low_voltage,
+                'user_connected': is_user_connected,
                 'signal_strength': rssi,
                 'reboot_counter': reboots,
                 'sequence': counter,
-                'broadcast_toggle': broadcast_toggle,
+                'broadcast_toggle': broadcast_toggle, # FIX toggle is not decrypted at this point
                 'timestamp': timestamp,
                 'battery': battery / 32.0,
                 'advertising_version':2}
 
+        key_type = AuthProvider.NoKey
+        if is_encrypted:
+            if is_device_key:
+                key_type = AuthProvider.DeviceKey
+            elif is_user_key:
+                key_type = AuthProvider.UserKey
+
+        if is_encrypted and (is_device_key or is_user_key):
+            try:
+                key = self._key_provider.get_rotated_key(key_type, device_id,
+                    reboot_counter=reboots,
+                    rotation_interval_power=EPHEMERAL_KEY_CYCLE_POWER,
+                    current_timestamp=timestamp)
+            except NotFoundError:
+                self._logger.warning("Key type {} is not found".format(key_type))
+                return info, timestamp, None, None, None, None, None
+
+            nonce = generate_nonce(device_id, timestamp, reboot_low, reboot_high_packed, counter_packed)
+
+            try:
+                decrypted_data = decrypt_payload(key, data[7:], nonce)
+            except ValueError:
```

Comment:
Should add or modify the logger line to include the exception. For the logger, `self._logger.warning("Advertising packet is not verified", exc_info=True)`

Reply:
Thank you, missed it